packages/marshal/src/types.js

```diff
@@ -2,70 +2,317 @@
+ * its primary.
+ *
+ * A Remotable is a PassableKey, which `sameKey` compares for euivalence
+ * according to its unforgeable identity, which effectively compares the
+ * unforgeable identities of their primaries.
+ *
+ * Remotables are unordered and cannot be sorted.
+ *
+ * A message eventually sent to a Remotable<T> is eventually delivered to
+ * its primary T in the home vat of the primary.
+ *
+ * A remotable is pass-by-remote. When a (primary or remote) remotable is
+ * passed as an argument in a message to an object in another vat, it arrives
+ * as a (primary or remote) remotable for the same primary.
+ *
+ * A Remotable can be a key in a WeakStore, i.e., it can be a key in
+ * a WeakMemoryMap or a WeakVirtualMap, or an element of a WeakMemorySet or
+ * a WeakVirtualSet. Note that there are no passable weak stores, since it
+ * doesn't make sense to pass something that cannot be enumerated.
+ */
+
+/**
+ * @typedef {Remotable | Promise} LeafCap
+ * The Corresponding PassStyles are "remotable" or "promise".
+ *
+ * Both remotables and promises may appear as the capability leaves of
+ * a tree of pass-by-copy data. Ideally, a passed promise should only be a
+ * promise for a passable, but we cannot know that at the time the promise is
+ * passed, so we cannot enforce that.
+ *
+ * Promises are not PassableKeys or Comparables, and so cannot be compared for
+ * equivalence. Promises are unordered and cannot be sorted.
+ *
+ * Promises have a lifecycle consisting of the following states
+ *   * Unresolved - no decision yet about what the promise designates.
+ *   * Resolved - decision made for this specific promise.
+ *      * Forwarded - delegate eventual designation to another promise.
+ *      * Settled - designation question permanently settled.
+ *         * Rejected - will never designate anything for this *reason*.
+ *         * Fulfilled - designates this non-thenable *fulfillment*.
+ *
+ * For local promises, this taxonomy applies without qualification. When
+ * promise p is forwareded to promise q, then when q fulfills to f, p
+ * is immediately identically fulfilled to f.
+ *
+ * A remote promise r can be modeled as a promise r in one vat that has been
+ * forwarded to a promise q in another vat. Once q is settled, given eventual
+ * connectivity and continued progress, r will eventually be settled. If
+ * q is rejected, r will eventually be rejected, either with the same reason
+ * or with a reason from the "system". If q is fulfulled, then r will
+ * eventually either be fulfilled with the result of passing the fulfullment
+ * (as if as an argument in a message), or r will be rejected with a reason
+ * from the "system". For example, if q's fulfillment is not passable, then
+ * r will be rejected with this non-passability as the reason. If q's
+ * fulfullment is a PrimaryRemotable T, then r must eventually fulfill to
+ * a Remote<T> with the same primary.
```

Comment:
Are these state transitions mirrored in a way that is "cached" on the other end?  We don't yet do this in Goblins, and I've had a longstanding TODO to look at it... eg, if machine/vat A has a presence of P which is a remote promise that eventually points at Bob, and another object does a `.then(...)` equivalent on P *after* A has already heard that P has resolved, A will not remember this and will ask to "subscribe/listen" to the remote promise again with a new resolver.  It would almost certainly be better to instead cache this information to prevent round trips.

(Presumably you are probably being smarter than I am yet!)

Reply:
Overall, the answer to your question is yes, the resolution on A is adequate and the `.then` will not cause further messages to be sent to B. However, this highlights some differences:

> a presence of P which is a remote promise

We need to arrive at common terminology. A presence is never a promise. For better or worse, we have used the term *presence* at the eventual-send / handled-promise layer of abstraction, whereas at the captp level we speak of *remotes*, which are a kind of *remotable*. Every captp remote is actually realized as a handled-promise *presence*.

Thus, presences and promises are disjoint. A remote promise fulfilled to a remote PrimaryRemotable is locally fulfilled to its corresponding Remote, which is a presence.

Whether the promise is resolved or not, a `.then` itself never causes a remote message to be sent. Rather, it just registers a local listener, the success and failure callbacks. If the remote promise is eventually fulfilled, then locally it will eventually be settled, either fulfilled to a local presence which is the remote, or rejected with an error. In both cases, all local `.then`s will eventually trigger.